In [59]:
# Véase http://dx.doi.org/10.1007/bf02840710
import math
import numpy as np
import random as rnd

In [60]:
iterations = 50
tolerancia = 1e-20

In [61]:
def promedios(Xs, Ys, N): 
    if (N % 2 == 1):
        first = Ys
        second = Xs
    else:
        first = Xs
        second = Ys

    
    for i in range(1,N):
        if (i % 2 == 1):
            first[i] = (first[i-1] + second[i-1])/2
        else:
            second[i] = (first[i-1] + second[i-1])/2

In [62]:
# Arreglos para guardar los valores de los promedios sucesivos
Xs = np.zeros(iterations)
Ys = np.zeros(iterations)
 
for N in range(1,iterations):
    sumaN = 0 # Acumulador para cada n y m posibles para determinado N
    
    for n in range(N+1): # n de 0 a N
        for m in range (n+1): # m no mayor a n
            dist = math.sqrt(N**2 + n**2 + m**2) # Distancia r_{ij}
            sign = (-1)**abs(N + n + m + 1) # Distancia r_{ij}

            # Casos mostrados en http://dx.doi.org/10.1007/bf02840710 p. 82
            if (n == 0 and m == 0):
                z = 6
            elif (N > 1 and m == 0 and n < N):
                z = 24
            elif (n == N and m == 0):
                z = 12
            elif (N > 2 and m < n and n < N):
                z = 48
            elif (N > 1 and m == n and n < N):
                z = 24
            elif (N > 1 and n == N and m < N):
                z = 24
            elif (n == m and N == n):
                z = 8

            sumaN = sumaN + z * dist**(-1) * sign

    if (N % 2 == 1): # Si N es impar, se guarda el nuevo dato en X. Si es par, en Y
        Xs[0] = sumaN + Ys[0]
    else: 
        Ys[0] = sumaN + Xs[0]

    # Promedios sucesivos para acelerar la convergencia
    promedios(Xs, Ys, N)

    if(N > 3 and abs(Xs[N-3] -  Ys[N-3]) < tolerancia):
        print("Se ha conseguido convergencia: ", rnd.choice([Xs[N-1],Ys[N-1]]))
        break

Se ha conseguido convergencia:  1.747564594633158


La constante de Mandelung para NaCl calculada es de: **1.747564594633158**